biz.aQute.launcher/src/aQute/launcher/Launcher.java

```diff
@@ -480,37 +480,29 @@ private int activate() throws Exception {
 		return result;
 	}
 
-	private boolean isAfterFrameworkStart(BundleActivator activator) {
+	private Object getImmediateFieldValue(BundleActivator activator) {
 		try {
 			Field f = activator.getClass()
 				.getField("IMMEDIATE");
-			if (f.getType()
-				.equals(Boolean.class)
-				|| f.getType()
-					.equals(boolean.class)) {
-				return false;
-			} else if (f.getType()
-				.equals(String.class)) {
-				return "AFTER_FRAMEWORK_START".equals(f.get(activator));
-			}
+			return f.get(activator);
 		} catch (Exception e) {}
-		return false;
+		return null;
+	}
+
+	private boolean isAfterFrameworkInit(BundleActivator activator) {
+		Object result = getImmediateFieldValue(activator);
+		return "AFTER_FRAMEWORK_INIT".equals(result);
 	}
 
 	private boolean isBeforeBundleStart(BundleActivator activator) {
-		try {
-			Field f = activator.getClass()
-				.getField("IMMEDIATE");
-			if (f.getType()
-				.equals(Boolean.class)
-				|| f.getType()
-					.equals(boolean.class)) {
-				return f.getBoolean(activator);
-			} else if (f.getType()
-				.equals(String.class)) {
-				return "BEFORE_BUNDLES_START".equals(f.get(activator));
+
+		Object result = getImmediateFieldValue(activator);
+		if (result != null) {
+			if (result instanceof String) {
+				return "BEFORE_BUNDLES_START".equals(result);
+			} else if (result instanceof Boolean || Boolean.TYPE.isInstance(result)) {
```

Comment:
You do not need the second check since the reflective call will never return a `boolean` since that cannot be represented as `Object`. It will return a `Boolean` object for a `boolean` field.

So you can implement `isBeforeBundlesStart` simply as:

```java
private boolean isBeforeBundlesStart(BundleActivator activator) {
  Object result = getImmediateFieldValue(activator);
  return "BEFORE_BUNDLES_START".equals(result) || Boolean.TRUE.equals(result);
}
```